File: WinForms/DevExpress.WindowsMailClient.Win/CS/DevExpress.WindowsMailClient.Win/Modules/Folders/MailViewControl/MailViewer.cs
using System;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;
using DevExpress.Data;
using DevExpress.Skins;
using DevExpress.Utils;
using DevExpress.Utils.MVVM;
using DevExpress.WindowsMailClient.Win.Data;
using DevExpress.WindowsMailClient.Win.Model;
using DevExpress.WindowsMailClient.Win.Utils;
using DevExpress.WindowsMailClient.Win.ViewModels;
using DevExpress.XtraBars;
using DevExpress.XtraBars.Navigation;
using DevExpress.XtraBars.ToastNotifications;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Controls;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Tile;
using DevExpress.XtraGrid.Views.Tile.ViewInfo;
using DevExpress.XtraLayout;
using Message = DevExpress.WindowsMailClient.Win.Data.Message;

namespace DevExpress.WindowsMailClient.Win.Modules {
    public enum FilterType {
        Focused,
        Other
    }
    public partial class MailViewer : BaseModule {
        Color UnreadTextColor;
        public MailViewer() : base(typeof(MailViewerViewModel)) {
            InitializeComponent();
            layoutControl1.UseLocalBindingContext = true;
            gridControl1.DataSource = MailClientDataModel.Messages;
            this.colDate.SortOrder = ColumnSortOrder.Descending;
            this.biSortDate.Tag = "Date";
            this.biSortFrom.Tag = "From";
            this.biSortRead.Tag = "Read";
            this.biSortSubject.Tag = "SubjectDisplayText";
            this.biSortImportance.Tag = "Priority";

            ViewModel.MessageDeleted += ViewModel_MessageDeleted;
            ViewModel.MessageFlagged += ViewModel_MessageFlagged;
            ViewModel.MessageChanged += ViewModel_MessageChanged;
            ViewModel.FilterChanged += ViewModel_FilterChanged;
            ViewModel.SortInfoChanged += ViewModel_SortInfoChanged;

            tileView.MouseMove += TileView_MouseMove;

            tileView.OptionsFind.AllowFindPanel = true;
            this.gridControl1.SizeChanged += (s, e) => { CompactTemplate = gridControl1.Width <= 300; };
            InitNotifications();

            ViewModel.CurrentFilterType = FilterType.Other;
            ViewModel.CurrentMailType = MailType.Inbox;

            navigationItemFocused.Tag = FilterType.Focused;
            navigationItemOther.Tag = FilterType.Other;

            BindCommands();
            InitBindings();

            UpdateColors();
        }

        void TileView_MouseMove(object sender, MouseEventArgs e) {
            if(tileView.IsEditing) return;
            TileViewHitInfo hitInfo = tileView.CalcHitInfo(e.X, e.Y) as TileViewHitInfo;
            if(hitInfo != null) currentRowHandle = hitInfo.RowHandle;
        }

        void ViewModel_MessageFlagged(object sender, EventArgs e) {
            SetMailType(ViewModel.CurrentMailType);
        }

        bool updateSortValues;
        void ViewModel_FilterChanged(object sender, EventArgs e) {
            UpdateCurrentMessage();

            updateSortValues = true;

            bcFilterAll.Checked = true;
            bcFilterUnread.Checked = false;
            bcFilterImportance.Checked = false;
            bcFilterToday.Checked = false;
            bcFilterYesterday.Checked = false;

            updateSortValues = false;
        }
        void UpdateCurrentMessage() {
            tileView.FocusedRowHandle = 0;
            ViewModel.CurrentMessage = tileView.GetFocusedRow() as Message;
        }
        public MailViewerViewModel ViewModel {
            get { return GetViewModel<MailViewerViewModel>(); }
        }
        void ViewModel_MessageChanged(object sender, EventArgs e) {
            mailPresenter.ShowMessage(ViewModel.CurrentMessage);
        }

        void InitBindings() {
            mvvmContext.SetBinding(tileView, view => view.ActiveFilterCriteria, "ActiveFilter");
            mvvmContext.SetBinding(labelHeader, x => x.Text, "LabelText");
            mvvmContext.SetBinding(lciHeader, x => x.Visibility, "LayoutHeaderVisibility");
            mvvmContext.SetBinding(lciTabFilter, x => x.Visibility, "FocusedItemVisibility");

            MVVMContextFluentAPI<MailViewerViewModel> fluentAPI = mvvmContext.OfType<MailViewerViewModel>();
            fluentAPI.EventToCommand<FocusedRowChangedEventArgs>(tileView, "FocusedRowChanged", x => x.SetCurrentMessage(null), (Func<FocusedRowChangedEventArgs, object>)GetMessageForFocusedRow);
            fluentAPI.EventToCommand<ItemClickEventArgs>(biSortDate, "ItemClick", x => x.Sort(null), EventArgsToSortCommandParameter());
            fluentAPI.EventToCommand<ItemClickEventArgs>(biSortFrom, "ItemClick", x => x.Sort(null), EventArgsToSortCommandParameter());
            fluentAPI.EventToCommand<ItemClickEventArgs>(biSortRead, "ItemClick", x => x.Sort(null), EventArgsToSortCommandParameter());
            fluentAPI.EventToCommand<ItemClickEventArgs>(biSortSubject, "ItemClick", x => x.Sort(null), EventArgsToSortCommandParameter());
            fluentAPI.EventToCommand<ItemClickEventArgs>(biSortImportance, "ItemClick", x => x.Sort(null), EventArgsToSortCommandParameter());

            fluentAPI.EventToCommand<ItemClickEventArgs>(bcFilterAll, "CheckedChanged", x => x.SetFilterAll(), x => ((BarCheckItem)x.Item).Checked && !updateSortValues);
            fluentAPI.EventToCommand<ItemClickEventArgs>(bcFilterUnread, "CheckedChanged", x => x.SetFilterUnread(), x => ((BarCheckItem)x.Item).Checked && !updateSortValues);
            fluentAPI.EventToCommand<ItemClickEventArgs>(bcFilterImportance, "CheckedChanged", x => x.SetFilterImportance(), x => ((BarCheckItem)x.Item).Checked && !updateSortValues);
            fluentAPI.EventToCommand<ItemClickEventArgs>(bcFilterToday, "CheckedChanged", x => x.SetFilterToday(), x => ((BarCheckItem)x.Item).Checked && !updateSortValues);
            fluentAPI.EventToCommand<ItemClickEventArgs>(bcFilterYesterday, "CheckedChanged", x => x.SetFilterYesterday(), x => ((BarCheckItem)x.Item).Checked && !updateSortValues);
            fluentAPI.EventToCommand<NavigationBarItemEventArgs>(officeFilterTab, "SelectedItemChanged", x => x.SetFilterType(GetCurrentFilterType()), x => GetCurrentFilterType());
        }
        object GetMessageForFocusedRow(FocusedRowChangedEventArgs x) {
            if(x.FocusedRowHandle >= 0) return (Message)tileView.GetFocusedRow();
            return null;
        }
        Func<ItemClickEventArgs, object> EventArgsToSortCommandParameter() {
            return x => {
                this.popupMenu.BeginUpdate();
                if(x.Item.ImageIndex == -1) {
                    UpdateSortItemsImageIndexes();
                    x.Item.ImageIndex = 0;
                }
                else {
                    x.Item.ImageIndex = x.Item.ImageIndex == 1 ? 0 : 1;
                }
                popupMenu.EndUpdate();
                return new SortInfo() { Column = tileView.Columns[(string)x.Item.Tag], Order = x.Item.ImageIndex == 1 ? ColumnSortOrder.Ascending : ColumnSortOrder.Descending };
            };
        }
        void ViewModel_MessageDeleted(object sender, EventArgs e) {
            SetMailType(ViewModel.CurrentMailType);
        }

        void BindCommands() {
            mvvmContext.BindCommand<MailViewerViewModel>(contextButton2, x => x.DeleteMessage(GetCurrentMessage()), p => GetCurrentMessage());
            mvvmContext.BindCommand<MailViewerViewModel>(contextButton3, x => x.ReplyMessage(GetCurrentMessage()), p => GetCurrentMessage());
            mvvmContext.BindCommand<MailViewerViewModel>(checkContextButton1, x => x.SetMessageFlaged(GetCurrentMessage()), p => GetCurrentMessage());
        }
        int currentRowHandle;
        Message GetCurrentMessage() { return tileView.GetRow(currentRowHandle) as Message; }
        FilterType GetCurrentFilterType() {
            var item = officeFilterTab.SelectedItem;
            return item != null ? (FilterType)item.Tag : FilterType.Other;
        }
        #region Notifications
        Timer notificationsTimer;
        XtraBars.Alerter.AlertControl alertControl;
        void InitNotifications() {
            if(CanUseToastNotifications()) {
                notificationManager.ApplicationId = AppProvider.ApplicationID;
                notificationManager.TryCreateApplicationShortcut();
                notificationManager.Activated += notificationsManager_Activated;
            }
            else {
                alertControl = new XtraBars.Alerter.AlertControl(components);
                alertControl.AllowHtmlText = true;
                alertControl.FormLocation = XtraBars.Alerter.AlertFormLocation.TopRight;
                alertControl.ShowPinButton = false;
                alertControl.AlertClick += alertControl_AlertClick;
            }
            EnsureNotificationsTimer();
        }
        void EnsureNotificationsTimer() {
            if(notificationsTimer == null) {
                notificationsTimer = new Timer(components);
                notificationsTimer.Interval = 15000;
                notificationsTimer.Tick += notificationsTimer_Tick;
            }
            notificationsTimer.Start();
        }
        void DestroyNotificationsTimer() {
            if(notificationsTimer != null) {
                notificationsTimer.Stop();
                notificationsTimer.Tick -= notificationsTimer_Tick;
                notificationsTimer.Dispose();
            }
            notificationsTimer = null;
        }
        void notificationsTimer_Tick(object sender, EventArgs e) {
            if(notificationManager.IsDisposing) {
                DestroyNotificationsTimer();
                return;
            }
            if(Utils.Utils.ShowNotifications) {
                notificationsTimer.Interval = 15000;
                ShowNotification();
            }
            
        }
        void alertControl_AlertClick(object sender, XtraBars.Alerter.AlertClickEventArgs e) {
            object notificationId = e.Info.Tag;
            e.AlertForm.Close();
            OnNotificationClick(notificationId);
        }
        void notificationsManager_Activated(object sender, ToastNotificationEventArgs e) {
            OnNotificationClick(e.NotificationID);
        }
        bool CanUseToastNotifications() {
            return ToastNotificationsManager.AreToastNotificationsSupported;
        }
        void ShowNotification() {
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
        }
        void OnNotificationClick(object notificationId) {
        }
        #endregion

        #region UI Customization and Drawing

        protected override void WndProc(ref System.Windows.Forms.Message m) {
            if(m.Msg == 528) { 
                this.GetParentViewModel<MainViewModel>().HideAllPanels();
            }
            base.WndProc(ref m);
        }
        void tileView1_CustomItemTemplate(object sender, TileViewCustomItemTemplateEventArgs e) {
            if(CompactTemplate)
                e.Template = e.Templates["smallsize"];
        }

        protected override void OnLookAndFeelChanged() {
            base.OnLookAndFeelChanged();
            UpdateColors();
        }

        void UpdateColors() {
            SkinElement skinElement = NavPaneSkins.GetSkin(LookAndFeel.ActiveLookAndFeel)[NavPaneSkins.SkinOfficeNavigationBarItem];
            UnreadTextColor = skinElement != null ? skinElement.Properties.GetColor("PressedColor") : ColorProvider.QuestionColor;
            this.tileView.Appearance.GroupText.ForeColor = UnreadTextColor;
            this.tileView.Appearance.ItemFocused.BackColor = Color.FromArgb(40, UnreadTextColor);
            this.tileView.Appearance.ItemHovered.BackColor = Color.FromArgb(40, UnreadTextColor);

            this.labelControl1.Appearance.Font =
            this.labelControl1.AppearanceHovered.Font =
            this.labelControl1.AppearancePressed.Font =
            this.labelControl1.AppearanceDropDown.Font =
            this.labelControl1.AppearanceDropDownHovered.Font = FontProvider.GetFont(labelControl1.Font.FontFamily.Name, 12f);

            this.labelControl1.Appearance.ForeColor =
            this.labelControl1.AppearanceHovered.ForeColor =
            this.labelControl1.AppearancePressed.ForeColor =
            this.labelControl1.AppearanceDropDown.ForeColor =
            this.labelControl1.AppearanceDropDownHovered.ForeColor =
            this.labelControl1.AppearanceDropDownPressed.ForeColor = UnreadTextColor;
        }
        void tileView_ItemCustomize(object sender, TileViewItemCustomizeEventArgs e) {
            bool isRead = (int)tileView.GetRowCellValue(e.RowHandle, colRead) == 1;
            if(!isRead) {
                e.Item["Read"].Appearance.Normal.BackColor = UnreadTextColor;
                e.Item["Date"].Appearance.Normal.ForeColor = UnreadTextColor;
                e.Item["SubjectDisplayText"].Appearance.Normal.ForeColor = UnreadTextColor;
                e.Item["SubjectDisplayText"].Appearance.Normal.FontStyleDelta = FontStyle.Bold;
            }
        }
        void tileView_ContextButtonCustomize(object sender, TileViewContextButtonCustomizeEventArgs e) {
            if(e.Item.Name == "cbFlag") {
                bool flagged = (int)tileView.GetRowCellValue(e.RowHandle, colFlag) % 2 != 0;
                e.Item.AllowGlyphSkinning = DefaultBoolean.False;
                e.Item.Visibility = flagged ? ContextItemVisibility.Visible : ContextItemVisibility.Auto;
            }
            else if(e.Item.Name == "cbExclamation") {
                bool priority = (int)tileView.GetRowCellValue(e.RowHandle, colPriority) == 2;
                e.Item.AllowGlyphSkinning = DefaultBoolean.False;
                e.Item.Visibility = priority ? ContextItemVisibility.Visible : ContextItemVisibility.Hidden;
            }
            else {
                e.Item.AllowGlyphSkinning = DefaultBoolean.True;
                e.Item.AppearanceHover.ForeColor = e.Item.AppearanceNormal.ForeColor = ((ITileControl)tileView.GetViewInfo()).ViewInfo.AppearanceText.ForeColor;
            }
        }
        #endregion

        bool _compactTemplate = false;
        bool CompactTemplate {
            get { return _compactTemplate; }
            set {
                if(_compactTemplate == value) return;
                _compactTemplate = value;
                tileView.RefreshData();
            }
        }
        void ViewModel_SortInfoChanged(object sender, EventArgs e) {
            tileView.SortInfo.Clear();
            tileView.SortInfo.Add(ViewModel.SortInfo);
        }
        void UpdateSortItemsImageIndexes() {
            this.biSortDate.ImageIndex = -1;
            this.biSortFrom.ImageIndex = -1;
            this.biSortRead.ImageIndex = -1;
            this.biSortSubject.ImageIndex = -1;
            this.biSortImportance.ImageIndex = -1;
        }
        public void SetMailType(MailType mailType) {
            ViewModel.CurrentMailType = mailType;
            this.gridControl1.RefreshDataSource();
        }
        protected override void OnDisposing() {
            ViewModel.MessageDeleted -= ViewModel_MessageDeleted;
            ViewModel.MessageFlagged -= ViewModel_MessageFlagged;
            ViewModel.MessageChanged -= ViewModel_MessageChanged;
            tileView.MouseMove -= TileView_MouseMove;
            base.OnDisposing();
        }

        void gridControl1_ControlAdded(object sender, ControlEventArgs e) {
            if(e.Control is FindControl) {
                if(e.Control.Controls.Count > 0) {
                    var layoutControls = e.Control.Controls.OfType<LayoutControl>();
                    foreach(var ctrl in layoutControls) {
                        var layoutControlGroup = ctrl.Items.FindByName("Root") as LayoutControlGroup;
                        if(layoutControlGroup != null) layoutControlGroup.Padding = new XtraLayout.Utils.Padding(18, 10, 5, 5);
                        var items = ctrl.Items.OfType<LayoutControlItem>();
                        foreach(LayoutControlItem item in items) {
                            item.Padding = new XtraLayout.Utils.Padding(item.Padding.Left, item.Padding.Right, 0, 0);
                        }
                    }
                }
            }
        }
    }
}

File: WinForms/DevExpress.WindowsMailClient.Win/VB/DevExpress.WindowsMailClient.Win/Modules/Folders/MailViewControl/MailViewer.vb
Imports System
Imports System.Drawing
Imports System.Linq
Imports System.Windows.Forms
Imports DevExpress.Data
Imports DevExpress.Skins
Imports DevExpress.Utils
Imports DevExpress.Utils.MVVM
Imports DevExpress.WindowsMailClient.Win.Data
Imports DevExpress.WindowsMailClient.Win.Model
Imports DevExpress.WindowsMailClient.Win.Utils
Imports DevExpress.WindowsMailClient.Win.ViewModels
Imports DevExpress.XtraBars
Imports DevExpress.XtraBars.Navigation
Imports DevExpress.XtraBars.ToastNotifications
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Controls
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Tile
Imports DevExpress.XtraGrid.Views.Tile.ViewInfo
Imports DevExpress.XtraLayout
Imports Message = DevExpress.WindowsMailClient.Win.Data.Message

Namespace DevExpress.WindowsMailClient.Win.Modules

    Public Enum FilterType
        Focused
        Other
    End Enum

    Public Partial Class MailViewer
        Inherits BaseModule

        Private UnreadTextColor As Color

        Public Sub New()
            MyBase.New(GetType(MailViewerViewModel))
            InitializeComponent()
            layoutControl1.UseLocalBindingContext = True
            gridControl1.DataSource = MailClientDataModel.Messages
            colDate.SortOrder = ColumnSortOrder.Descending
            biSortDate.Tag = "Date"
            biSortFrom.Tag = "From"
            biSortRead.Tag = "Read"
            biSortSubject.Tag = "SubjectDisplayText"
            biSortImportance.Tag = "Priority"
            AddHandler ViewModel.MessageDeleted, AddressOf ViewModel_MessageDeleted
            AddHandler ViewModel.MessageFlagged, AddressOf ViewModel_MessageFlagged
            AddHandler ViewModel.MessageChanged, AddressOf ViewModel_MessageChanged
            AddHandler ViewModel.FilterChanged, AddressOf ViewModel_FilterChanged
            AddHandler ViewModel.SortInfoChanged, AddressOf ViewModel_SortInfoChanged
            AddHandler tileView.MouseMove, AddressOf TileView_MouseMove
            tileView.OptionsFind.AllowFindPanel = True
            AddHandler gridControl1.SizeChanged, Sub(s, e) CompactTemplate = gridControl1.Width <= 300
            InitNotifications()
            ViewModel.CurrentFilterType = FilterType.Other
            ViewModel.CurrentMailType = MailType.Inbox
            navigationItemFocused.Tag = FilterType.Focused
            navigationItemOther.Tag = FilterType.Other
            BindCommands()
            InitBindings()
            UpdateColors()
        End Sub

        Private Sub TileView_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
            If tileView.IsEditing Then Return
            Dim hitInfo As TileViewHitInfo = TryCast(tileView.CalcHitInfo(e.X, e.Y), TileViewHitInfo)
            If hitInfo IsNot Nothing Then currentRowHandle = hitInfo.RowHandle
        End Sub

        Private Sub ViewModel_MessageFlagged(ByVal sender As Object, ByVal e As EventArgs)
            SetMailType(ViewModel.CurrentMailType)
        End Sub

        Private updateSortValues As Boolean

        Private Sub ViewModel_FilterChanged(ByVal sender As Object, ByVal e As EventArgs)
            UpdateCurrentMessage()
            updateSortValues = True
            bcFilterAll.Checked = True
            bcFilterUnread.Checked = False
            bcFilterImportance.Checked = False
            bcFilterToday.Checked = False
            bcFilterYesterday.Checked = False
            updateSortValues = False
        End Sub

        Private Sub UpdateCurrentMessage()
            tileView.FocusedRowHandle = 0
            ViewModel.CurrentMessage = TryCast(tileView.GetFocusedRow(), Message)
        End Sub

        Public ReadOnly Property ViewModel As MailViewerViewModel
            Get
                Return GetViewModel(Of MailViewerViewModel)()
            End Get
        End Property

        Private Sub ViewModel_MessageChanged(ByVal sender As Object, ByVal e As EventArgs)
            mailPresenter.ShowMessage(ViewModel.CurrentMessage)
        End Sub

        Private Sub InitBindings()
            mvvmContext.SetBinding(tileView, Function(view) view.ActiveFilterCriteria, "ActiveFilter")
            mvvmContext.SetBinding(labelHeader, Function(x) x.Text, "LabelText")
            mvvmContext.SetBinding(lciHeader, Function(x) x.Visibility, "LayoutHeaderVisibility")
            mvvmContext.SetBinding(lciTabFilter, Function(x) x.Visibility, "FocusedItemVisibility")
            Dim fluentAPI As MVVMContextFluentAPI(Of MailViewerViewModel) = mvvmContext.OfType(Of MailViewerViewModel)()
            fluentAPI.EventToCommand(tileView, "FocusedRowChanged", Sub(x) x.SetCurrentMessage(Nothing), CType(AddressOf GetMessageForFocusedRow, Func(Of FocusedRowChangedEventArgs, Object)))
            fluentAPI.EventToCommand(biSortDate, "ItemClick", Sub(x) x.Sort(Nothing), EventArgsToSortCommandParameter())
            fluentAPI.EventToCommand(biSortFrom, "ItemClick", Sub(x) x.Sort(Nothing), EventArgsToSortCommandParameter())
            fluentAPI.EventToCommand(biSortRead, "ItemClick", Sub(x) x.Sort(Nothing), EventArgsToSortCommandParameter())
            fluentAPI.EventToCommand(biSortSubject, "ItemClick", Sub(x) x.Sort(Nothing), EventArgsToSortCommandParameter())
            fluentAPI.EventToCommand(biSortImportance, "ItemClick", Sub(x) x.Sort(Nothing), EventArgsToSortCommandParameter())
            fluentAPI.EventToCommand(Of ItemClickEventArgs)(bcFilterAll, "CheckedChanged", Sub(x) x.SetFilterAll(), Function(x) CType(x.Item, BarCheckItem).Checked AndAlso Not updateSortValues)
            fluentAPI.EventToCommand(Of ItemClickEventArgs)(bcFilterUnread, "CheckedChanged", Sub(x) x.SetFilterUnread(), Function(x) CType(x.Item, BarCheckItem).Checked AndAlso Not updateSortValues)
            fluentAPI.EventToCommand(Of ItemClickEventArgs)(bcFilterImportance, "CheckedChanged", Sub(x) x.SetFilterImportance(), Function(x) CType(x.Item, BarCheckItem).Checked AndAlso Not updateSortValues)
            fluentAPI.EventToCommand(Of ItemClickEventArgs)(bcFilterToday, "CheckedChanged", Sub(x) x.SetFilterToday(), Function(x) CType(x.Item, BarCheckItem).Checked AndAlso Not updateSortValues)
            fluentAPI.EventToCommand(Of ItemClickEventArgs)(bcFilterYesterday, "CheckedChanged", Sub(x) x.SetFilterYesterday(), Function(x) CType(x.Item, BarCheckItem).Checked AndAlso Not updateSortValues)
            fluentAPI.EventToCommand(Of NavigationBarItemEventArgs)(officeFilterTab, "SelectedItemChanged", Sub(x) x.SetFilterType(GetCurrentFilterType()), Function(x) GetCurrentFilterType())
        End Sub

        Private Function GetMessageForFocusedRow(ByVal x As FocusedRowChangedEventArgs) As Object
            If x.FocusedRowHandle >= 0 Then Return CType(tileView.GetFocusedRow(), Message)
            Return Nothing
        End Function

        Private Function EventArgsToSortCommandParameter() As Func(Of ItemClickEventArgs, Object)
            Return Function(x)
                popupMenu.BeginUpdate()
                If x.Item.ImageIndex = -1 Then
                    UpdateSortItemsImageIndexes()
                    x.Item.ImageIndex = 0
                Else
                    x.Item.ImageIndex = If(x.Item.ImageIndex = 1, 0, 1)
                End If

                popupMenu.EndUpdate()
                Return New SortInfo() With {.Column = tileView.Columns(CStr(x.Item.Tag)), .Order = If(x.Item.ImageIndex = 1, ColumnSortOrder.Ascending, ColumnSortOrder.Descending)}
            End Function
        End Function

        Private Sub ViewModel_MessageDeleted(ByVal sender As Object, ByVal e As EventArgs)
            SetMailType(ViewModel.CurrentMailType)
        End Sub

        Private Sub BindCommands()
            mvvmContext.BindCommand(Of MailViewerViewModel)(contextButton2, Sub(x) x.DeleteMessage(GetCurrentMessage()), Function(p) GetCurrentMessage())
            mvvmContext.BindCommand(Of MailViewerViewModel)(contextButton3, Sub(x) x.ReplyMessage(GetCurrentMessage()), Function(p) GetCurrentMessage())
            mvvmContext.BindCommand(Of MailViewerViewModel)(checkContextButton1, Sub(x) x.SetMessageFlaged(GetCurrentMessage()), Function(p) GetCurrentMessage())
        End Sub

        Private currentRowHandle As Integer

        Private Function GetCurrentMessage() As Message
            Return TryCast(tileView.GetRow(currentRowHandle), Message)
        End Function

        Private Function GetCurrentFilterType() As FilterType
            Dim item = officeFilterTab.SelectedItem
            Return If(item IsNot Nothing, CType(item.Tag, FilterType), FilterType.Other)
        End Function

#Region "Notifications"
        Private notificationsTimer As Timer

        Private alertControl As XtraBars.Alerter.AlertControl

        Private Sub InitNotifications()
            If CanUseToastNotifications() Then
                notificationManager.ApplicationId = AppProvider.ApplicationID
                notificationManager.TryCreateApplicationShortcut()
                AddHandler notificationManager.Activated, AddressOf notificationsManager_Activated
            Else
                alertControl = New Alerter.AlertControl(components)
                alertControl.AllowHtmlText = True
                alertControl.FormLocation = Alerter.AlertFormLocation.TopRight
                alertControl.ShowPinButton = False
                AddHandler alertControl.AlertClick, AddressOf alertControl_AlertClick
            End If

            EnsureNotificationsTimer()
        End Sub

        Private Sub EnsureNotificationsTimer()
            If notificationsTimer Is Nothing Then
                notificationsTimer = New Timer(components)
                notificationsTimer.Interval = 15000
                AddHandler notificationsTimer.Tick, AddressOf notificationsTimer_Tick
            End If

            notificationsTimer.Start()
        End Sub

        Private Sub DestroyNotificationsTimer()
            If notificationsTimer IsNot Nothing Then
                notificationsTimer.Stop()
                RemoveHandler notificationsTimer.Tick, AddressOf notificationsTimer_Tick
                notificationsTimer.Dispose()
            End If

            notificationsTimer = Nothing
        End Sub

        Private Sub notificationsTimer_Tick(ByVal sender As Object, ByVal e As EventArgs)
            If notificationManager.IsDisposing Then
                DestroyNotificationsTimer()
                Return
            End If

            If ShowNotifications Then
                notificationsTimer.Interval = 15000
                ShowNotification()
            End If
        End Sub

        Private Sub alertControl_AlertClick(ByVal sender As Object, ByVal e As XtraBars.Alerter.AlertClickEventArgs)
            Dim notificationId As Object = e.Info.Tag
            e.AlertForm.Close()
            OnNotificationClick(notificationId)
        End Sub

        Private Sub notificationsManager_Activated(ByVal sender As Object, ByVal e As ToastNotificationEventArgs)
            OnNotificationClick(e.NotificationID)
        End Sub

        Private Function CanUseToastNotifications() As Boolean
            Return ToastNotificationsManager.AreToastNotificationsSupported
        End Function

        Private Sub ShowNotification()
        End Sub

        Private Sub OnNotificationClick(ByVal notificationId As Object)
        End Sub

#End Region
#Region "UI Customization and Drawing"
        Protected Overrides Sub WndProc(ByRef m As Windows.Forms.Message)
            If m.Msg = 528 Then
                Me.GetParentViewModel(Of MainViewModel)().HideAllPanels()
            End If

            MyBase.WndProc(m)
        End Sub

        Private Sub tileView1_CustomItemTemplate(ByVal sender As Object, ByVal e As TileViewCustomItemTemplateEventArgs)
            If CompactTemplate Then e.Template = e.Templates("smallsize")
        End Sub

        Protected Overrides Sub OnLookAndFeelChanged()
            MyBase.OnLookAndFeelChanged()
            UpdateColors()
        End Sub

        Private Sub UpdateColors()
            Dim skinElement As SkinElement = NavPaneSkins.GetSkin(LookAndFeel.ActiveLookAndFeel)(NavPaneSkins.SkinOfficeNavigationBarItem)
            UnreadTextColor = If(skinElement IsNot Nothing, skinElement.Properties.GetColor("PressedColor"), QuestionColor)
            tileView.Appearance.GroupText.ForeColor = UnreadTextColor
            tileView.Appearance.ItemFocused.BackColor = Color.FromArgb(40, UnreadTextColor)
            tileView.Appearance.ItemHovered.BackColor = Color.FromArgb(40, UnreadTextColor)
            labelControl1.AppearanceDropDownHovered.Font = GetFont(labelControl1.Font.FontFamily.Name, 12F)
            labelControl1.AppearanceDropDown.Font = labelControl1.AppearanceDropDownHovered.Font
            labelControl1.AppearancePressed.Font = labelControl1.AppearanceDropDown.Font
            labelControl1.AppearanceHovered.Font = labelControl1.AppearancePressed.Font
            labelControl1.Appearance.Font = labelControl1.AppearanceHovered.Font
            labelControl1.AppearanceDropDownPressed.ForeColor = UnreadTextColor
            labelControl1.AppearanceDropDownHovered.ForeColor = labelControl1.AppearanceDropDownPressed.ForeColor
            labelControl1.AppearanceDropDown.ForeColor = labelControl1.AppearanceDropDownHovered.ForeColor
            labelControl1.AppearancePressed.ForeColor = labelControl1.AppearanceDropDown.ForeColor
            labelControl1.AppearanceHovered.ForeColor = labelControl1.AppearancePressed.ForeColor
            labelControl1.Appearance.ForeColor = labelControl1.AppearanceHovered.ForeColor
        End Sub

        Private Sub tileView_ItemCustomize(ByVal sender As Object, ByVal e As TileViewItemCustomizeEventArgs)
            Dim isRead As Boolean = CInt(tileView.GetRowCellValue(e.RowHandle, colRead)) = 1
            If Not isRead Then
                e.Item("Read").Appearance.Normal.BackColor = UnreadTextColor
                e.Item("Date").Appearance.Normal.ForeColor = UnreadTextColor
                e.Item("SubjectDisplayText").Appearance.Normal.ForeColor = UnreadTextColor
                e.Item("SubjectDisplayText").Appearance.Normal.FontStyleDelta = FontStyle.Bold
            End If
        End Sub

        Private Sub tileView_ContextButtonCustomize(ByVal sender As Object, ByVal e As TileViewContextButtonCustomizeEventArgs)
            If Equals(e.Item.Name, "cbFlag") Then
                Dim flagged As Boolean = CInt(tileView.GetRowCellValue(e.RowHandle, colFlag)) Mod 2 <> 0
                e.Item.AllowGlyphSkinning = DefaultBoolean.False
                e.Item.Visibility = If(flagged, ContextItemVisibility.Visible, ContextItemVisibility.Auto)
            ElseIf Equals(e.Item.Name, "cbExclamation") Then
                Dim priority As Boolean = CInt(tileView.GetRowCellValue(e.RowHandle, colPriority)) = 2
                e.Item.AllowGlyphSkinning = DefaultBoolean.False
                e.Item.Visibility = If(priority, ContextItemVisibility.Visible, ContextItemVisibility.Hidden)
            Else
                e.Item.AllowGlyphSkinning = DefaultBoolean.True
                e.Item.AppearanceNormal.ForeColor = CType(tileView.GetViewInfo(), ITileControl).ViewInfo.AppearanceText.ForeColor
                e.Item.AppearanceHover.ForeColor = e.Item.AppearanceNormal.ForeColor
            End If
        End Sub

#End Region
        Private _compactTemplate As Boolean = False

        Private Property CompactTemplate As Boolean
            Get
                Return _compactTemplate
            End Get

            Set(ByVal value As Boolean)
                If _compactTemplate = value Then Return
                _compactTemplate = value
                tileView.RefreshData()
            End Set
        End Property

        Private Sub ViewModel_SortInfoChanged(ByVal sender As Object, ByVal e As EventArgs)
            tileView.SortInfo.Clear()
            tileView.SortInfo.Add(ViewModel.SortInfo)
        End Sub

        Private Sub UpdateSortItemsImageIndexes()
            biSortDate.ImageIndex = -1
            biSortFrom.ImageIndex = -1
            biSortRead.ImageIndex = -1
            biSortSubject.ImageIndex = -1
            biSortImportance.ImageIndex = -1
        End Sub

        Public Sub SetMailType(ByVal mailType As MailType)
            ViewModel.CurrentMailType = mailType
            gridControl1.RefreshDataSource()
        End Sub

        Protected Overrides Sub OnDisposing()
            RemoveHandler ViewModel.MessageDeleted, AddressOf ViewModel_MessageDeleted
            RemoveHandler ViewModel.MessageFlagged, AddressOf ViewModel_MessageFlagged
            RemoveHandler ViewModel.MessageChanged, AddressOf ViewModel_MessageChanged
            RemoveHandler tileView.MouseMove, AddressOf TileView_MouseMove
            MyBase.OnDisposing()
        End Sub

        Private Sub gridControl1_ControlAdded(ByVal sender As Object, ByVal e As ControlEventArgs)
            If TypeOf e.Control Is FindControl Then
                If e.Control.Controls.Count > 0 Then
                    Dim layoutControls = e.Control.Controls.OfType(Of LayoutControl)()
                    For Each ctrl In layoutControls
                        Dim layoutControlGroup = TryCast(ctrl.Items.FindByName("Root"), LayoutControlGroup)
                        If layoutControlGroup IsNot Nothing Then layoutControlGroup.Padding = New XtraLayout.Utils.Padding(18, 10, 5, 5)
                        Dim items = ctrl.Items.OfType(Of LayoutControlItem)()
                        For Each item As LayoutControlItem In items
                            item.Padding = New XtraLayout.Utils.Padding(item.Padding.Left, item.Padding.Right, 0, 0)
                        Next
                    Next
                End If
            End If
        End Sub
    End Class
End Namespace
